### PR TITLE
printing.preview: Use the windows API to find a default viewer application

### PR DESCRIPTION
**Pull request by eric-wieser**
#### Brief description of what is fixed or changed

A new `sympy.printing.preview.system_default_viewer` constant, which can be passed as the `viewer` argument to `sympy.preview`.

On windows, this is now the default viewer for non-png formats.

On all platforms where it is installed, `pyglet` continues to be the default viewer for png.

On non-windows platforms, there is a hard-coded list of viewers for `dvi`, `ps`, and `pdf`.
If none of these are found, `system_default_viewer` is used.

#### Other comments

This has been tested only on windows.

#### Release Notes

* printing
  * `preview` now supports a `viewer=system_default_viewer` argument, which is now the default on windows and used as a final fallback on all platforms.

### CODE REVIEW

File: sympy/printing/preview.py [hunk 1]
```diff
@@ -29,15 +29,100 @@ def _check_output_no_window(*args, **kwargs):
...
+def system_default_viewer(fname, fmt):
+    import platform
+    if platform.system() == 'Darwin':
+        import subprocess
+        subprocess.call(('open', fname))
+    elif platform.system() == 'Windows':
+        _start_file(fname)
+    else:
+        import subprocess
+        subprocess.call(('xdg-open', fname))
```
Comment (oscarbenjamin) [inline sympy/printing/preview.py:111]: On some Linux systems this would be `gio open fname`

Comment (eric-wieser) [reply sympy/printing/preview.py:111]: Which systems? Presumably `xdg-open` does not exist in those systems

Comment (oscarbenjamin) [reply sympy/printing/preview.py:111]: Looks like maybe I got confused by the same warning discussed here:
https://github.com/github/hub/issues/1473

**Comment by asmeurer** [comment]
Is the default viewer still pyglet? I think we should use the system viewer by default, especially on Mac and Linux where I the system viewer function will always work.

**Comment by eric-wieser** [comment]
The default viewer for PNG is pyglet on all platforms, but now only if it is installed.

**Comment by asmeurer** [comment]
Should we add `viewer='system'` as a shorthand for this?

**Comment by asmeurer** [comment]
I think at least for Mac and presumably Windows we should make the system viewer the default. pyglet is very janky, especially on Mac.

By the way, I have some issue with my preview, where it just produces empty images. This happens for both pyglet and the system viewer. I think maybe the temporary image is being deleted too soon (I get a lot of "this file doesn't exist" errors when using Preview). 

Also, it would be a good idea to use the `str` form of the input expression somehow for the filename, rather than just `texput.png`.

**Comment by asmeurer** [comment]
I guess you've seen https://github.com/sympy/sympy/pull/13621/. It would be good to get those fixes in. The docstring of preview isn't very good (even with that PR), in that it is very confusing what the options are and what their defaults are. We should add a `Parameters` section so it is clearer.

**Comment by eric-wieser** [comment]
> . I think maybe the temporary image is being deleted too soon (I get a lot of "this file doesn't exist" errors when using Preview).

I had to go to great lengths to avoid this problem on windows. Does xdg-open block until the application is done?

**Comment by oscarbenjamin** [comment]
The `open` program/command on OSX does not appear to block (I just tested it). There is a `-W` argument to block which did seem to work:
```
$ open --help
open: unrecognized option `--help'
Usage: open [-e] [-t] [-f] [-W] [-R] [-n] [-g] [-h] [-s <partial SDK name>][-b <bundle identifier>] [-a <application>] [filenames] [--args arguments]
Help: Open opens files from a shell.
      By default, opens each file using the default application for that file.  
      If the file is in the form of a URL, the file will be opened as a URL.
Options: 
      -a                Opens with the specified application.
      -b                Opens with the specified application bundle identifier.
      -e                Opens with TextEdit.
      -t                Opens with default text editor.
      -f                Reads input from standard input and opens with TextEdit.
      -F  --fresh       Launches the app fresh, that is, without restoring windows. Saved persistent state is lost, excluding Untitled documents.
      -R, --reveal      Selects in the Finder instead of opening.
      -W, --wait-apps   Blocks until the used applications are closed (even if they were already running).
          --args        All remaining arguments are passed in argv to the application's main() function instead of opened.
      -n, --new         Open a new instance of the application even if one is already running.
      -j, --hide        Launches the app hidden.
      -g, --background  Does not bring the application to the foreground.
      -h, --header      Searches header file locations for headers matching the given filenames, and opens them.
      -s                For -h, the SDK to use; if supplied, only SDKs whose names contain the argument value are searched.
                        Otherwise the highest versioned SDK in each platform is used.
```
This all seems like too much platform specific code for sympy to me. Does anyone even use preview? (Presumably not many people because the pyglet version is seriously buggy and isn't a required dependency)

I didn't realise that startfile was Windows-only. A platform independent version of startfile seems like a great idea for the stdlib but I don't think it belongs in sympy.

**Comment by asmeurer** [comment]
I think a lot of people use preview(). There was a post on the mailing list just the other day where someone was using it. Platform specific code is unfortunately unavoidable for a function like this. We also have platform specific code in other places line autowrap and Unicode support detection in the printers. 

Making open block is not ideal because the user would need to quit preview to continue using sympy.

**Comment by oscarbenjamin** [comment]
I think that using matplotlib is preferable to having platform-specific code: https://github.com/sympy/sympy/issues/8446#issuecomment-557798066. SymPy already has an enormous scope just dealing with "all of mathematics": we should let other libraries deal with the vagaries of different OSes.

**Comment by eric-wieser** [comment]
> Making open block is not ideal because the user would need to quit preview to continue using sympy.

Maybe we should just not delete the file from under the viewer then, and leave it it temp for the OS to clean up?

**Comment by eric-wieser** [comment]
> Should we add `viewer='system'` as a shorthand for this?

I'm not a fan of this - what if someone has a `system` binary on their machine which they want to use as a viewer? We should be using out-of-band values for things which aren't binary names.

Perhaps a more convenient spelling would be `preview(..., viewer=preview.system)`, since actually importing the full name is rather challenging.

**Comment by eric-wieser** [comment]
> Making open block is not ideal because the user would need to quit preview to continue using sympy.

I think this is already the behavior today for the non-system viewers anyway.

**Comment by eric-wieser** [comment]
> There was a post on the mailing list just the other day where someone was using it

I wasn't able to find that - do you have a link?

**Comment by oscarbenjamin** [comment]
> If we wanted to change the behavior of `preview`, we could choose not to delete the file from under the viewer, and leave it it temp for the OS to clean up. That would eliminate the `_start_file` function.

That seems reasonable to me.

**Comment by asmeurer** [comment]
> I wasn't able to find that - do you have a link?

Ah, I was confusing the post where the person was using `display()` (I'm not sure what that is). 

I do think preview is used, though. I use it myself occasionally to get a png image of an expression. Actually, using the system viewer would make this work even better because instead of passing a filename, I could just use preview() and then Cmd-S in Preview.

> Perhaps a more convenient spelling would be preview(..., viewer=preview.system), since actually importing the full name is rather challenging.

IMO the preview() API is already too complicated, i.e., to save an expression to a file or bytesio requires passing two arguments when it could easily just do one. So I would prefer to keep things simple even at the expense of disallowing certain executable names (which aren't likely to occur anyway).

> If we wanted to change the behavior of preview, we could choose not to delete the file from under the viewer, and leave it it temp for the OS to clean up. That would eliminate the _start_file function.

We should definitely do that. The OS should (presumably) be smart enough to not delete temp files that are currently open.

**Comment by asmeurer** [comment]
Regarding blocking with `open` on macOS, let's see what the behavior is like when we use an OS tempfile. If we don't have to block that would be preferable. 

Regarding matplotlib, it's not the greatest image viewer. I like using the system viewer as the default. it is designed to view images. And it easily enables the use-case of saving an expression as an image, which I think is one of the most useful features of preview().

**Comment by eric-wieser** [comment]
> We should definitely do that. The OS should (presumably) be smart enough to not delete temp files that are currently open.

This is tricky, because the file isn't open yet at the point python is done with it

**Comment by eric-wieser** [comment]
> the person was using `display()` (I'm not sure what that is).

It's an ipython built-in that calls `_repr_latex_` or `_repr_html_` when in jupyter notebooks.

**Comment by czgdp1807** [comment]
@eric-wieser Is this PR still on? Please resolve the conflicts if it is. Thank you.

**Comment by czgdp1807** [comment]
@eric-wieser The tests are failing. Please let us know if this PR can be concluded or if it can be take over by other contributors. Thank you.

**Comment by eric-wieser** [comment]
Test failure looks unrelated, it's a timeout in an ode test

**Comment by oscarbenjamin** [comment]
Yes, I think the failure is #20456. I've restarted the tests.

**Comment by oscarbenjamin** [comment]
I just tried `preview(x+1)` on OSX and I get a small blank white window on master and with this PR.

I tried this with the PR:
```
In [1]: from sympy.printing.preview import system_default_viewer                                                                  

In [2]: preview(x**2 + 1, viewer=system_default_viewer)
```
The first time nothing happened.

Second time I got a message from "Preview" saying something like "Unable to open texput.png because there is no such file".

Third time I got a large Preview Window that was all black and that I couldn't close.

Fourth time it closed the black Window I couldn't close and gave me a new Window that was all white (including the top bar where the close button should be).

I can quite preview with cmd-q but otherwise I'm not getting anything useful from the system viewer here.

**Comment by eric-wieser** [comment]
@oscarbenjamin, does adjusting `open` to pass `-W` behave correctly?

File: sympy/printing/preview.py [hunk 2]
```diff
@@ -27,9 +27,94 @@ def _check_output_no_window(*args, **kwargs):
...
+        WaitForSingleObject(proc, -1)
+    finally:
+        CloseHandle(proc)
+
+
+def system_default_viewer(fname, fmt):
+    import platform
+    if platform.system() == 'Darwin':
+        import subprocess
+        subprocess.call(('open', fname))
```
Comment (eric-wieser) [inline sympy/printing/preview.py:104]:
```suggestion
        # ensure the process starts before the temp file is removed
        subprocess.call(('open', '-W', fname))
```

**Comment by oscarbenjamin** [comment]
> does adjusting `open` to pass `-W` behave correctly?

Yes, although it blocks until I quit the Preview application. If I just close the preview window then the terminal is still blocked: I have to `Cmd-Q` preview which is not ideal because it closes all other Preview windows (e.g. Preview is the default PDF viewer).

**Comment by eric-wieser** [comment]
> I have to `Cmd-Q` preview which is not ideal because it closes all other Preview windows

Yeah, that's clearly not acceptable behavior. Last test - what about `-W` and `-n`?

File: sympy/printing/preview.py (hunk 2 above)
Comment (eric-wieser) [reply sympy/printing/preview.py:104]: Or
```suggestion
        # ensure the process starts before the temp file is removed, but don't require all windows of the viewer to
        # be closed.
        subprocess.call(('open', '-W', '-n', fname))
```

**Comment by oscarbenjamin** [comment]
With `-W` and `-n` I still have to quit preview.

If I quit with cmd-q then it doesn't close any other preview windows and it does unblock the terminal which is what we want but that's not the normal way to close a window.

If I close the window with cmd-w or click the close button then the Window goes but the terminal doesn't unblock. Worse if I switch desktops then the application quits secretly but still doesn't unblock the terminal. Then I can't even quit it and the only way is to ctrl-c the terminal.

**Comment by eric-wieser** [comment]
To be clear - is the behavior worse with this PR, or just no better? If it's not any worse than it was before on OSX, then I'd argue this is still worth putting in to make windows users happy, and to follow up to fix OSX later; but if it's worse, then I guess we need to fix them all at once.

**Comment by oscarbenjamin** [comment]
The behaviour is not worse. The default is that preview doesn't work for me either way. I thought I'd test out the new viewer here to see how it works but it does not seem to work on OSX. Maybe that's just this computer though (I don't have other OSX machines to test).

Longer term I do think it's worth considering potentially more robust approaches though e.g.:

1. Use Mathjax (doesn't require a local latex installation) and open the result in a browser.

2. Use matplotlib to display the image because it already handles blocking, and should be able to work in notebooks etc.

3. Use a browser to view the image (could serve on localhost and then not worry about temporary files).

I'm happy to merge this as a fix for Windows but I'd be happier if I could actually test it myself!

Can anyone else try this out to see how it works?

**Comment by oscarbenjamin** [comment]
It would be great if someone could test this on Windows and report the results from e.g.:
```
>>> from sympy import *
>>> x = Symbol('x')
>>> preview(1 + x**2)
```
I don't have easy access to Windows but I'm confident that @eric-wieser has tested this reasonably. Sometimes differences in systems can be overlooked though and it's good to maintain the principle of independent testing where possible.

In the absence of feedback I'll just merge this without testing but I don't think that's good practice.

**Comment by eric-wieser** [comment]
Just tested again - `output="pdf"` combined with `sumatraPDF` works for me, but using the windows photo viewer with `output="png"` crashes.

**Comment by eric-wieser** [comment]
So maybe this isn't a suitable approach after all, and we instead just need to not delete the temporary files, as suggested earlier.

**Comment by eric-wieser** [comment]
Alright, I've updated this to just ensure the file isn't deleted, so that the system viewer doesn't have any issues.

**Comment by oscarbenjamin** [comment]
> I tried this with the PR:
> 
> ```
> In [1]: from sympy.printing.preview import system_default_viewer                                                                  
> 
> In [2]: preview(x**2 + 1, viewer=system_default_viewer)
> ```

I just retried this after the latest commit and it works great (on OSX). No blocking and I can see the equation. Maybe this should be the default on OSX because the default shows nothing for me.

Can anyone else on OSX confirm whether this or the default preview works for them?

Also can anyone confirm that this PR works on Windows?

**Comment by eric-wieser** [comment]
> Maybe this should be the default on OSX because the default shows nothing for me.

I assume the default for you is to use `pyglet`, and that doesn't work?

File: sympy/printing/preview.py [hunk 3]
```diff
@@ -122,21 +149,24 @@ def preview(expr, output='png', viewer=None, euler=True, packages=(),
...
-    You may also enter 'file' for the viewer argument. Doing so will cause
-    this function to return a file object in read-only mode, if 'filename'
+    You may also enter ``'file'`` for the viewer argument. Doing so will cause
+    this function to return a file object in read-only mode, if ``filename``
     is unset. However, if it was set, then 'preview' writes the genereted
     file to this filename instead.
 
-    There is also support for writing to a BytesIO like object, which needs
-    to be passed to the 'outputbuffer' argument.
+    There is also support for writing to a :class:`BytesIO` like object, which
```
Comment (eric-wieser) [inline sympy/printing/preview.py:168]:
```suggestion
    There is also support for writing to a :class:`io.BytesIO` like object, which
```

File: sympy/printing/preview.py [hunk 4]
```diff
@@ -154,63 +184,62 @@ def preview(expr, output='png', viewer=None, euler=True, packages=(),
     If the value of 'output' is different from 'dvi' then command line
     options can be set ('dvioptions' argument) for the execution of the
     'dvi'+output conversion tool. These options have to be in the form of a
-    list of strings (see subprocess.Popen).
+    list of strings (see :any:`subprocess.Popen`).
 
-    Additional keyword args will be passed to the latex call, e.g., the
-    symbol_names flag.
+    Additional keyword args will be passed to the :func:`sympy.latex` call,
```
Comment (eric-wieser) [inline sympy/printing/preview.py:189]:
```suggestion
    Additional keyword args will be passed to the :func:`~sympy.printing.latex.latex` call,
```

File: sympy/printing/preview.py [hunk 5]
```diff
@@ -154,63 +184,62 @@ def preview(expr, output='png', viewer=None, euler=True, packages=(),
     If the value of 'output' is different from 'dvi' then command line
     options can be set ('dvioptions' argument) for the execution of the
     'dvi'+output conversion tool. These options have to be in the form of a
-    list of strings (see subprocess.Popen).
+    list of strings (see :any:`subprocess.Popen`).
```
Comment (eric-wieser) [inline sympy/printing/preview.py:187]:
```suggestion
    list of strings (see :class:`subprocess.Popen`).
```

File: sympy/printing/preview.py [hunk 6]
```diff
@@ -122,21 +149,24 @@ def preview(expr, output='png', viewer=None, euler=True, packages=(),
...
-    You may also enter 'file' for the viewer argument. Doing so will cause
-    this function to return a file object in read-only mode, if 'filename'
+    You may also enter ``'file'`` for the viewer argument. Doing so will cause
+    this function to return a file object in read-only mode, if ``filename``
     is unset. However, if it was set, then 'preview' writes the genereted
     file to this filename instead.
 
-    There is also support for writing to a BytesIO like object, which needs
-    to be passed to the 'outputbuffer' argument.
+    There is also support for writing to a :class:`io.BytesIO` like object, which
```
Comment (eric-wieser) [inline sympy/printing/preview.py:168]:
```suggestion
    There is also support for writing to a ``io.BytesIO`` like object, which
```

File: sympy/printing/preview.py [hunk 7]
```diff
@@ -154,63 +184,62 @@ def preview(expr, output='png', viewer=None, euler=True, packages=(),
     If the value of 'output' is different from 'dvi' then command line
     options can be set ('dvioptions' argument) for the execution of the
     'dvi'+output conversion tool. These options have to be in the form of a
-    list of strings (see subprocess.Popen).
+    list of strings (see :class:`subprocess.Popen`).
```
Comment (eric-wieser) [inline sympy/printing/preview.py:187]:
```suggestion
    list of strings (see ``subprocess.Popen``).
```

**Review by eric-wieser — COMMENTED** [review]
Looks like intersphinx is not enabled on the sympy docs, so links to python builtins don't work,

**Comment by oscarbenjamin** [comment]
There is an issue for intersphinx in the sympy docs: https://github.com/sympy/sympy/issues/19788

Do you know how to make it work?

We could fix this when updating the docs for 1.7.1.

**Comment by oscarbenjamin** [comment]
> > Maybe this should be the default on OSX because the default shows nothing for me.
> 
> I assume the default for you is to use `pyglet`, and that doesn't work?

Seems so. I tried uninstalling pyglet but then I get:
```
ImportError: pyglet is required for preview.
 visit http://www.pyglet.org/
```

**Comment by eric-wieser** [comment]
> you know how to make it work?

Well, I already fixed that bug in the PR linked to in the conversation there :)

That issue is about using interpsphinx to link _to_ sympy; my remark about was about linking _from_ sympy.

Do you know which of the two conf.py files is used to generate the sympy docs?

**Comment by oscarbenjamin** [comment]
The sphinx build failed on Travis but passed on actions. I've restarted it on Travis. Hopefully it's just a sporadic failure. Otherwise maybe the actions build isn't checking this properly.

**Comment by eric-wieser** [comment]
I suspect the error was real - it's often quite hard to find in the sphinx logs, looking for `WARNING` usually finds it.

**Comment by oscarbenjamin** [comment]
> > you know how to make it work?
> 
> Well, I already fixed that bug in the PR linked to in the conversation there :)

Can that issue be closed then?

> That issue is about using interpsphinx to link _to_ sympy; my remark about was about linking _from_ sympy.

Okay, that makes sense.

> Do you know which of the two conf.py files is used to generate the sympy docs?

The answer is probably no but I'm also not even sure I understand the question. I only see one .conf file:
```
$ find . -name '*.conf'
./doc/apidoc.conf
```

**Comment by oscarbenjamin** [comment]
> I suspect the error was real - it's often quite hard to find in the sphinx logs, looking for `WARNING` usually finds it.

If it is a real error then the actions test for sphinx needs to be fixed because it didn't find any error. Hopefully we are going to switch off Travis soon so we need to make sure that actions is functioning as our replacement CI.

**Comment by oscarbenjamin** [comment]
It was a real fail:
```
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:1376: MatplotlibDeprecationWarning: The 's' parameter of annotate() has been renamed 'text' since Matplotlib 3.3; support for the old name will be dropped two minor releases later.
  ax.annotate(**a)
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:1501: RuntimeWarning: invalid value encountered in double_scalars
  cos_theta = dot_product / (vector_a_norm * vector_b_norm)
/home/travis/build/sympy/sympy/sympy/plotting/plot.py:1501: RuntimeWarning: invalid value encountered in double_scalars
  cos_theta = dot_product / (vector_a_norm * vector_b_norm)
/home/travis/build/sympy/sympy/sympy/printing/preview.py:docstring of sympy.printing.preview.preview:69: WARNING: py:class reference target not found: subprocess.Popen
generating indices... genindex py-modindex done
writing additional pages... search done
copying images... [  1%] modules/matrices/../../../_build/plot_directive/modules/matrices/matrices-1.png
...
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 1 warning.
Makefile:49: recipe for target 'html' failed
make: *** [html]
```

**Comment by oscarbenjamin** [comment]
I've opened a PR to fix the sphinx test on actions https://github.com/sympy/sympy/pull/20581

**Comment by oscarbenjamin** [comment]
Closing and reopening to restart the tests

**Comment by oscargus** [comment]
I think this would be nice to get in for 1.11. It is basically what @asmeurer asked me to add to #22734

Would you have time to rebase @eric-wieser ?

**Comment by eric-wieser** [comment]
I'll have a look today

**Comment by oscarbenjamin** [comment]
This looks good to me but isn't something that I can test. Can someone review/test this?

**Comment by eric-wieser** [comment]
I thought I removed the code that deleted the temporary file?

File: sympy/printing/preview.py [hunk 8]
```diff
@@ -25,10 +25,38 @@ def _check_output_no_window(*args, **kwargs):
...
+    """ Open fname with the default system viewer.
+
+    In practice, it is impossible for python to know when the system viewer is
+    done. For this reason, we ensure the passed file will not be deleted under
+    it, and this function does not attempt to block.
+    """
+    # copy to a new temporary file that will not be deleted
+    with tempfile.NamedTemporaryFile(prefix='sympy-preview-',
+                                     suffix=os.path.splitext(fname)[1],
+                                     delete=False) as temp_f:
```
Comment (eric-wieser) [inline sympy/printing/preview.py:38]: @asmeurer: here's where I request it not be deleted

**Comment by asmeurer** [comment]
Oh my bad. I didn't notice that when I did a 'pull' that it didn't work (I guess I still had your remote as a `git://` URL). The latest version of this branch does indeed seem to work (the default resolution of the image could be improved, but that's a separate issue). 

It also does still try to use pyglet when it's installed, which IMO it shouldn't (at least on Mac, where pyglet has never worked correctly for me). 

I am not able to test Linux or Windows presently.

**Comment by eric-wieser** [comment]
> It also does still try to use pyglet when it's installed

I was working on the assumption that I should keep that behavior since it was the old default; and that it would make sense to revisit that decision separately.

**Comment by eric-wieser** [comment]
> The only potential thing that possibly can be improved is to delete the temporary file on Linux where it is typically possible. But I do not really see that as major issue (and possibly a source of major issues on some dists?).

This feels impossible to me; I don't think `xdg-open` is specified to block, and if someone has configured `image/png` to open in a tabbed interface like `chrome`, then probably `chrome my_file.png` sends a message to a running chrome instance, then exits before waiting for a full page load at the other end.

**Comment by oscargus** [comment]
You are correct. I didn't think of that option.

**Comment by eric-wieser** [comment]
I suppose it would be possible to configure python to delete the file at exit, although that would still be annoying for people running a python script that ends in `preview()`, as the interpreter could still exit before the viewer opens.

**Comment by oscarbenjamin** [comment]
> I don't think `xdg-open` is specified to block

Actually this works nicely on Linux if I uninstall pyglet. If I have pyglet installed then I get a small window like the one that Aaron showed except the lower half of the window is black. When I uninstall pyglet then on master preview bails with "pyglet is requires for preview" but with this PR I can see something sensible.

**Comment by oscarbenjamin** [comment]
> I don't think that's related to the part of my comment you quoted?

Yeah, sorry. That message got changed as I wrote it. I was going to begin by saying that it does seem to block but it's actually pyglet that blocks. When I uninstall pyglet and use this PR then preview works without blocking.

**Comment by oscarbenjamin** [comment]
If I've understood the comments above correctly then this has been tested on Windows (by @oscargus and @eric-wieser) and on OSX (by @asmeurer) and on Linux by me. In all cases this was found to be an improvement when pyglet is not installed. It's probably hard to get more comprehensive testing without just releasing this and seeing what happens.

A separate issue to follow up can be just removing the use of pyglet altogether but that is out of scope in this particular PR. Probably most users don't have pyglet so this will just be an occasional trip up for some people but obviously it would ultimately be good to remove this.

I'm going to merge this now. Since the 1.11 release branch is now created this would need to be backported with an explicit PR against the 1.11 branch for it to be included in the 1.11 release.

**Comment by oscarbenjamin** [comment]
Thanks @eric-wieser !

**Pull request merged**